### PR TITLE
Fixes #20925 - k-change-hostname should check exit codes

### PR DESCRIPTION
**Pull request by johnpmitsch**
We should check the exit codes on each command ran in the script and exit if they fail.

### CODE REVIEW

File: katello/katello-change-hostname [hunk 1]
```diff
@@ -256,13 +255,14 @@ unless @foreman_proxy_content
   STDOUT.puts "\nUpdating default #{@proxy}"
   default_capsule_id = hammer_cmd("capsule list | grep #{old_hostname} | awk '{ print $1 }'").chomp
   # Incorrect error message is piped to /dev/null, can be removed when http://projects.theforeman.org/issues/18186 is fixed
-  hammer_cmd("capsule update --id #{default_capsule_id} --url https://#{@new_hostname}:9090 --new-name #{@new_hostname} 2> /dev/null")
+  # For the same reason, we can't use the run_cmd method here
+  hammer_cmd("capsule update --id #{default_capsule_id} --url https://#{@new_hostname}:9090 --new-name #{@new_hostname} 2> /dev/null", [0, 65])
```
Comment (johnpmitsch) [inline katello/katello-change-hostname]: ugg.... this is hacky. This command always errors due to the issue mentioned in the comments above, so we pipe the ouput to `/dev/null/`. I added an exception for the exit code it returns, but this is assuming the exit code is always the same. 

@evgeni @akofink Is there a better way to handle this? And if not, can I rely on the exit code always being the same on this error? This approach worked in my testing.

The tl;dr of [the bug](http://projects.theforeman.org/issues/18186) is that it will update the capsule successfully but hammer returns an error.

**Comment by evgeni** [comment]
` run_cmd("hammer ping", [0], "There is a problem with the server, please check 'hammer ping'")`

this will fail on a proxy. the old code had a unless block around it.

**Comment by evgeni** [comment]
you can add a `|| true` at the end of the command. also hacky. but will always exit 0.

**Comment by johnpmitsch** [comment]
@evgeni thanks, I still have to test on a foreman-proxy!

**Comment by evgeni** [comment]
unrelated, but just realized, is the port really always 9090? shouldn't we read the old port and re use it?

**Comment by evgeni** [comment]
the puppet server/master restart will fail, as it's either puppetserver or httpd (for puppet3) which we have to restart.

**Comment by johnpmitsch** [comment]
> unrelated, but just realized, is the port really always 9090? shouldn't we read the old port and re use it?

@evgeni I'm not sure about the 9090 port, @ehelms or @jlsherrill do you know?

**Comment by johnpmitsch** [comment]
> the puppet server/master restart will fail, as it's either puppetserver or httpd (for puppet3) which we have to restart.

The original cause of adding those lines is [here](http://projects.theforeman.org/issues/18270/). In what scenario is this going to fail?

**Comment by evgeni** [comment]
if you have a Puppet3 setup, the puppet master is running inside Apache httpd using mod_passenger. With Puppet 4 you have the stand-alone puppetserver process. depending on that you have to restart different services. and with run_cmd this will now really fail and wont be silently discarded on a puppet 3 install which has no puppetserver.

**Comment by ehelms** [comment]
I think its 99% safe to assume smart proxies are running on 9090 in Katello deployments.

File: katello/katello-change-hostname [hunk 2]
```diff
@@ -166,13 +167,13 @@ def precheck
...
-  unless hammer_ping_success
-    fail_with_message("There is a problem with the server, please check 'hammer ping'")
+  unless @foreman_proxy_content
+    STDOUT.puts "\nChecking overall health of server"
+    run_cmd("hammer ping", [0], "There is a problem with the server, please check 'hammer ping'")
   end
 
   STDOUT.puts "\nChecking credentials"
-  hammer_cmd("capsule list")
+  hammer_cmd("capsule list") unless @foreman_proxy_content
```
Comment (evgeni) [inline katello/katello-change-hostname:176]: hah, good catch!

File: katello/katello-change-hostname (hunk 1 above)
Comment (evgeni) [reply katello/katello-change-hostname]: So we could do
```
hammer_cmd("capsule update --id #{default_capsule_id} --url https://#{@new_hostname}:9090 --new-name #{@new_hostname} 2> /dev/null || true")
```
here, or keep the 65 exit code as ok.

I think actually keeping 65 is better, as that limits the possibly-accepted errors (and should be stable: https://github.com/theforeman/hammer-cli/blob/master/lib/hammer_cli/exit_codes.rb#L5).

Just for me to understand: it thinks the new URL is invalid, as at the point you are executing this, the service was not yet really moved, right?

**Review by evgeni — APPROVED** [review]
ACK if my understanding in https://github.com/Katello/katello-packaging/pull/530/files#r139149900 is correct

**Comment by johnpmitsch** [comment]
@evgeni I'm not sure of the exact cause of updating the capsule w/ hammer working but throwing an error. Its a bad bug and we should fix soon! In my experience, it is consistent. I've tried out of the script too.
